File: Source/Server/Game/GameLogic.vb
Imports Core

Module GameLogic
    Function GetTotalMapPlayers(MapNum As Integer) As Integer
        Dim i As Integer, n As Integer
        n = 0

        For i = 1 To Socket.HighIndex()
            If GetPlayerMap(i) = mapNum Then
                n = n + 1
            End If
        Next

        GetTotalMapPlayers = n
    End Function

    Function GetNpcMaxVital(NpcNum As Integer, Vital As [Enum].VitalType) As Integer
        ' Prevent subscript out of range
        If NpcNum <= 0 Or NpcNum > MAX_NPCS Then Exit Function

        Select Case Vital
            Case VitalType.HP
                GetNpcMaxVital = Type.NPC(NPCNum).HP
            Case VitalType.SP
                GetNpcMaxVital = Type.NPC(NPCNum).Stat(StatType.Intelligence) * 2
            Case VitalType.SP
                GetNpcMaxVital = Type.NPC(NPCNum).Stat(StatType.Spirit) * 2
        End Select

    End Function

    Function FindPlayer(Name As String) As Integer
        Dim i As Integer

        For i = 1 To Socket.HighIndex()
            ' Trim and convert both names to uppercase for case-insensitive comparison
            If UCase$(GetPlayerName(i)) = UCase$(Name) Then
                FindPlayer = i
                Exit Function
            End If
        Next

        FindPlayer = 0
    End Function

    Friend Function CheckGrammar(Word As String, Optional Caps As Byte = 0) As String
        Dim FirstLetter As String

        FirstLetter = LCase$(Left$(Word, 1))

        If FirstLetter = "$" Then
            CheckGrammar = (Mid$(Word, 2, Len(Word) - 1))
            Exit Function
        End If

        If FirstLetter Like "*[aeiou]*" Then
            If Caps Then CheckGrammar = "An " & Word Else CheckGrammar = "an " & Word
        Else
            If Caps Then CheckGrammar = "A " & Word Else CheckGrammar = "a " & Word
        End If
    End Function

End Module

File: Source/Server/Game/GameLogic.cs
using Microsoft.VisualBasic;
using Microsoft.VisualBasic.CompilerServices;
using static Core.Global.Command;

namespace Server
{

    public class GameLogic
    {
        public static int GetTotalMapPlayers(int mapNum)
        {
            int GetTotalMapPlayersRet = default;
            int i;
            int n;
            n = 0;

            var loopTo = NetworkConfig.Socket.HighIndex;
            for (i = 0; i <= loopTo; i++)
            {
                if (GetPlayerMap(i) == mapNum)
                {
                    n = n + 1;
                }
            }

            GetTotalMapPlayersRet = n;
            return GetTotalMapPlayersRet;
        }

        public static int GetNpcMaxVital(double NpcNum, Core.Vital Vital)
        {
            int GetNpcMaxVitalRet = default;
            // Prevent subscript out of range
            if (NpcNum < 0 | NpcNum > Core.Constant.MAX_NPCS)
                return GetNpcMaxVitalRet;

            switch (Vital)
            {
                case Core.Vital.Health:
                    {
                        GetNpcMaxVitalRet = Core.Data.Npc[(int)NpcNum].HP;
                        break;
                    }
                case Core.Vital.Stamina:
                    {
                        GetNpcMaxVitalRet = (int)Core.Data.Npc[(int)NpcNum].Stat[(byte)Core.Stat.Intelligence] * 2;
                        break;
                    }
            }

            return GetNpcMaxVitalRet;

        }

        public static int FindPlayer(string Name)
        {
            int FindPlayerRet = default;
            int i;

            var loopTo = NetworkConfig.Socket.HighIndex;
            for (i = 0; i <= loopTo; i++)
            {
                // Trim and convert both names to uppercase for case-insensitive comparison
                if (Strings.UCase(GetPlayerName(i)) == Strings.UCase(Name))
                {
                    FindPlayerRet = i;
                    return FindPlayerRet;
                }
            }

            FindPlayerRet = -1;
            return FindPlayerRet;
        }

        public static string CheckGrammar(string Word, byte Caps = 0)
        {
            string CheckGrammarRet = default;
            string FirstLetter;

            FirstLetter = Strings.LCase(Strings.Left(Word, 1));

            if (FirstLetter == "$")
            {
                CheckGrammarRet = Strings.Mid(Word, 2, Strings.Len(Word) - 1);
                return CheckGrammarRet;
            }

            if (LikeOperator.LikeString(FirstLetter, "*[aeiou]*", CompareMethod.Binary))
            {
                if (Conversions.ToBoolean(Caps))
                    CheckGrammarRet = "An " + Word;
                else
                    CheckGrammarRet = "an " + Word;
            }
            else if (Conversions.ToBoolean(Caps))
                CheckGrammarRet = "A " + Word;
            else
                CheckGrammarRet = "a " + Word;
            return CheckGrammarRet;
        }

    }
}